File: MachineIntelligence/e6/e6-1b.py
from __future__ import print_function

import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout
from keras.optimizers import adam
from keras.initializers import random_normal, constant

batch_size = 128
num_classes = 10
epochs = 20

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

x_train = x_train.reshape(60000, 784)
x_test = x_test.reshape(10000, 784)
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
# print(x_train.shape[0], 'train samples')
# print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

model = Sequential()
model.add(Dense(units = 1500, activation = 'relu', input_shape = (784, ), 
    kernel_initializer = random_normal(mean = 0, stddev = 0.01), bias_initializer = constant(0.1)))
model.add(Dense(units = 1500, activation = 'relu', 
    kernel_initializer = random_normal(mean = 0, stddev = 0.01), bias_initializer = constant(0.1)))
model.add(Dense(units = 1500, activation = 'relu',
    kernel_initializer = random_normal(mean = 0, stddev = 0.01), bias_initializer = constant(0.1)))
model.add(Dense(units = 10, activation = 'softmax'))

# model.summary()

model.compile(
    loss = 'categorical_crossentropy',
    optimizer = adam(lr = 0.001, beta_1 = 0.001, beta_2 = 0.999, epsilon = 10e-8),
    metrics = ['accuracy']
)

history = model.fit(x_train, y_train,
    batch_size= 100, epochs= 10
)

print(history)

File: MachineIntelligence/e6/e6-1b.R
library(keras)
library(ggplot2)

script.dir <- dirname(sys.frame(1)$ofile)
setwd(script.dir)

mnist <- dataset_mnist()

# reshape
x_train <- array_reshape(mnist$train$x, c(nrow(mnist$train$x), 784))
x_test <- array_reshape(mnist$test$x, c(nrow(mnist$test$x), 784))
# rescale
x_train <- x_train / 255
x_test <- x_test / 255

y_train <- to_categorical(mnist$train$y, 10)
y_test <- to_categorical(mnist$test$y, 10)

model <- keras_model_sequential()

model %>% layer_dense(units = 1500, activation = "relu", input_shape = c(784))

model %>% 
    layer_dense(units = 1500, activation = "relu", input_shape = c(784)) %>%
    layer_dense(units = 1500, activation = "relu") %>%
    layer_dense(units = 1500, activation = "relu") %>%
    layer_dense(units = 10, activation = "softmax")

# C:\\Users\\IBM_ADMIN\\Desktop\\supreme-bassoon\\MachineIntelligence\\e6

model %>% compile(
  loss = 'categorical_crossentropy',
  optimizer = optimizer_adam(lr = 0.001, beta_1 = 0.001, beta_2 = 0.999, epsilon = 10e-8),
  metrics = c('accuracy')
)

history <- model %>% fit(
  x_train, y_train,
  epochs = 10, batch_size = 100,
)

plt.df <- data.frame(
  x = c(1:length(history$metrics$loss)),
  loss = history$metrics$loss,
  acc = history$metrics$acc
)

ggplot(plt.df, aes(x = x)) + geom_point(aes(y = loss, color = "Loss")) + geom_point(aes(y = acc, color = "Acc")) + labs(color = "Type")

ggsave("e6-1b.png")